### DISCUSSION TITLE
Compiler error does not help me figure out the problem with this code:

### DISCUSSION THREAD

**Discussion by andreweades** [pointfreeco/swift-parsing#155] [Q&A] [answered]
Each time I try to replace my current parsers with swift-parsing I get thwarted by compiler errors almost immediately. It's not at all obvious to me how to fix them though.

Here is an example of what I'm struggling with. No matter what I try, I cannot get it to compile so I must be missing something fundamental.

```
import Parsing

enum StackElement {
  case number(Double)
  case string(String)
}

let stackElement = Parse {
// ----------------^
// Error: "Type of expression is ambiguous without more context"
  StackElement.number($0)
} with: {
  Int.parser().map { Double($0)! }
}
```

**Comment by mbrandonw** [comment]
I'm going to convert this to a discussion and answer over there.

**Comment by mbrandonw** [comment, marked as answer]
There's a few things going on here that are causing bad error messaging. First of all, `Double`'s initializer is not failable so you can drop the force unwrap:

```diff
 let stackElement = Parse {
   StackElement.number($0)
 } with: {
-  Int.parser().map { Double($0)! }
+  Int.parser().map { Double($0) }
 }
```

Now the error message is "Ambiguous use of 'init(_:)'". This is happening because `Int.parser()` is overloaded quite a bit in order to support parsing on multiple string representations such as `Substring` and `UTF8View`, and in the current situation it can't figure out the type of input. You can explicitly specify the input to fix it:

```diff
 let stackElement = Parse {
   StackElement.number($0)
 } with: {
-  Int.parser().map { Double($0) }
+  Int.parser(of: Substring.self).map { Double($0) }
 }
```

And now it compiles.

It's worth mentioning that you will only run into this ambiguity problem when using `Int.parser()` in isolation, which from our experience is rare. Usually it's used in conjunction with other parsers that help Swift figure out the types, e.g.:

```swift
try Parse { 
  Int.parser()
  "!"
}
.parse("42!") // 42
```

Hope this helps!

**Comment by andreweades** [comment]
Thanks. I was trying everything including the errant “!” to get a successful compile. But this is a super helpful explanation.

My existing parsers use a Parseable protocol for the inputs so I have a ParseableString type which uses Substring. I think this is why I struggle to convert my parsers so your suggestion should put that right.
